### DISCUSSION TITLE
Ability to write multiple RowSets to Parquet files

### DISCUSSION THREAD

**Discussion by skh1000** [duckdb/duckdb#3870] [Ideas]
This is a suggestion as to how to "append to an existing Parquet file" which is currently impossible --
but, with proper support, you could just leave a parquet file writer open and append another rowset to it,
before closing it with a separate call.
This is my request, if anyone is willing to work on it (sorry, I'm not able to) --
basically you would 
1) write a table to a parquet file with a "leave file open" flag,
and then could 
2) write to the same file (perhaps even restricted to the same table) using an "append to file" flag, and perhaps a "leave open" flag to allow further appends,
 and then 
3) close the file explicitly when done (or do a (perhaps null) write without the "leave open" flag, 
which would then close the file).

This would allow writing very large parquet files, without having to have everything in memory first, and without the kludge of writing separate parquet files and treating them (on read) as one.

If anyone wants to work on this, let me know, it would be very helpful.

**Comment by Mytherin** [comment]
We already support streaming writes to Parquet, so there is no need for everything to fit in memory first (as long as you can express everything in a single query). The Parquet format is not really suitable for appending as the metadata of the Parquet file lives in the end of the file, so after you have written a Parquet file "that is it". You cannot easily append to an existing Parquet file like you can with e.g. CSV, as the format is simply not designed for this. 

If you want to incrementally build up a Parquet file using a procedural language I would perhaps look at Apache Arrow or similar projects, as SQL is not really suitable for this style of programming to begin with.

**Comment by skh1000** [comment]
You're the expert, so I shouldn't argue too much, but it's a shame, since the use case is so close.  Just to be clear, I'm not considering adding to an existing Parquet file, but writing it in "chunks" which would be done via separate calls to the SQL engine (and if I understand correctly, Parquet does have "chunks" which are row groups or whatever they're called).   I think the problem is that it's weird to keep a file open in SQL, you really want things organized around a "transaction" or a single operation, which may be long-lived, but is otherwise constrained.  But I'm not sure I agree that SQL isn't good for this -- SQL is great at providing a type scheme, and row definitions as a combination of types, and has benefits in terms of integration in that you don't have to include a lot of extra libraries and such -- ODBC (as with any database) should be sufficient.

Can you think of (or tell me) any other way that a parquet file could be written in "chunks" using your system?   I.e. the chunks should be collected / passed via separate calls to the SQL engine.  But also, if there were a single transaction / operation writing the parquet file as a stream with a long-time-open connection streaming in data (like, as an insert), and the process with the connection sent chunks at different times (Microsoft has a way to do this, the chunking of an insert stream, but I'm not sure about ODBC), then that would suffice.  I guess I also don't care what the chunk size or boundaries is/are, if that were handled independently of what my "chunks" were, that would be fine as well.

Again, the reason to do this is that the entire data might not fit in the database before hand, so it doesn't pre-exist in its entirety in a table or collection of tables.

The alternative is (as you say) to write code that's specific to the particular format, instead of using a general facility.  Also, we've had trouble integrating the huge code base and support libraries necessary to write parquet files using (I think) Arrow, and doing it instead via ODBC sounds very appealing.

**Reply by Mytherin** [reply]
While it would be technically feasible to do this using DuckDB, we would need to come up with a syntax to support this type of operation, as it is quite unusual to batch this type of writing in SQL to an external format over multiple queries. If you are working on this as part of a commercial enterprise, perhaps you would be interested in contracting [DuckDB Labs](https://duckdblabs.com) to implement this feature with you in co-operation.

**Comment by skh1000** [comment]
Yes.

**Comment by skh1000** [comment]
I can think of ways to do this that might be preferable, but I'll leave that to discuss with DuckDB Labs when they contact me.
